File: DataStreamasDisjointIntervals2.cpp
#include<iostream>
#include<vector>

using namespace std;

struct Interval {
	int start;
	int end;
	Interval() : start(0), end(0) {}
	Interval(int s, int e) : start(s), end(e) {}
};

class SummaryRanges {
private:
	vector<Interval> intervals;
	
	int binarySearch(vector<Interval> intervals, int l, int r, int val) {
		int mid;
		while(l <= r) {
			mid = l + (r - l) / 2;
			if(intervals[mid].end < val) l = mid + 1;
			else if(intervals[mid].start > val) r = mid - 1;
			else return mid;
		}
		return r;
	}
public:
    /** Initialize your data structure here. */
    SummaryRanges() {
    }
    
    void addNum(int val) {
        if(intervals.empty()) intervals.push_back(Interval(val, val));
        else {
        	int index = binarySearch(intervals, 0, intervals.size() - 1, val);
        	if(index == -1) {
        		if(val == intervals[0].start - 1) intervals[0].start--;
        		else if(val < intervals[0].start) intervals.insert(intervals.begin(), Interval(val, val));
			}else if(index == intervals.size() - 1) {
				if(val == intervals.back().end + 1) intervals[intervals.size() - 1].end++;
				else if(val > intervals.back().end) intervals.push_back(Interval(val, val));
			}else {
				if(val == intervals[index].end + 1 && val == intervals[index + 1].start - 1) {
					intervals[index].end = intervals[index + 1].end;
					intervals.erase(intervals.begin() + index + 1);
				}else if(val == intervals[index].end + 1) intervals[index].end++;
				else if(val == intervals[index + 1].start - 1) intervals[index + 1].start--;
				else if(val > intervals[index].end && val < intervals[index + 1].start) intervals.insert(intervals.begin() + index + 1, Interval(val, val));
			}
		}
    }
    
    vector<Interval> getIntervals() {
        return this->intervals;
    }
};

int main() {
	
	SummaryRanges *obj = new SummaryRanges();
	for(int i = 0; i < 5; i++) {
		int val;
		cin>>val;
		obj->addNum(val);
		vector<Interval> params = obj->getIntervals();
		for(int j = 0; j < params.size(); j++) {
			cout<<params[j].start<<"  "<<params[j].end<<endl;
		}
	}
	
	return 0;
	
}

File: DataStreamasDisjointIntervals2.py
class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e


class SummaryRanges:

    class BSTNode:
        def __init__(self, val):
            self.interval = val
            self.left = None
            self.right = None

    def __init__(self):
        self.root = None

    def addNum(self, val):
        """
        :type val: int
        :rtype: void
        """
        def findMin(root):
            if not root:
                return None
            if not root.left:
                return root
            return findMin(root.left)

        def remove(root, interval):
            if not root:
                return None
            if interval.start > root.interval.end:
                root.right = remove(root.right, interval)
            elif interval.end < root.interval.start:
                root.left = remove(root.left, interval)
            else:
                if root.left and root.right:
                    root.interval = findMin(root.right).interval
                    root.right = remove(root.right, root.interval)
                else:
                    root = root.left if root.left else root.right
            return root

        def findKey(root, val):
            if not root:
                return None
            elif val > root.interval.end:
                return findKey(root.right, val)
            elif val < root.interval.start:
                return findKey(root.left, val)
            else:
                return root

        def addNum(root, val):
            if not root:
                root = self.BSTNode(Interval(val, val))
            elif val > root.interval.end:
                root.right = addNum(root.right, val)
            elif val < root.interval.start:
                root.left = addNum(root.left, val)
            return root

        if not self.root:
            self.root = addNum(self.root, val)
        else:
            if findKey(self.root, val):
                return
            left = findKey(self.root, val - 1)
            right = findKey(self.root, val + 1)
            if not left and not right:
                self.root = addNum(self.root, val)
            elif left and not right:
                left.interval.end += 1
            elif not left and right:
                right.interval.start -= 1
            else:
                e = right.interval.end
                self.root = remove(self.root, right.interval)
                left.interval.end = e

    def getIntervals(self):
        """
        :rtype: List[Interval]
        """
        def inOrder(root):
            if not root:
                return
            inOrder(root.left)
            intervals.append(root.interval)
            inOrder(root.right)

        intervals = []
        inOrder(self.root)
        return intervals


if __name__ == '__main__':
    summary = SummaryRanges()
    num = int(input())
    while num != -1:
        summary.addNum(num)
        for interval in summary.getIntervals():
            print("%d  %d" % (interval.start, interval.end))
        num = int(input())
